src/libraries/System.Net.NameResolution/tests/FunctionalTests/MetricsTest.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Linq;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.NameResolution.Tests
+{
+    public class MetricsTest
+    {
+        private const string DnsLookupsRequested = "dns-lookups-requested";
+
+        [Fact]
+        public static async Task ResolveValidHostName_MetricsRecorded()
+        {
+            const string ValidHostName = "localhost";
+
+            using var recorder = new InstrumentRecorder<long>(DnsLookupsRequested);
+
+            await Dns.GetHostEntryAsync(ValidHostName);
+            await Dns.GetHostAddressesAsync(ValidHostName);
+
+            Dns.GetHostEntry(ValidHostName);
+            Dns.GetHostAddresses(ValidHostName);
+
+            Dns.EndGetHostEntry(Dns.BeginGetHostEntry(ValidHostName, null, null));
+            Dns.EndGetHostAddresses(Dns.BeginGetHostAddresses(ValidHostName, null, null));
+
+            long[] measurements = GetMeasurementsForHostname(recorder, ValidHostName);
+
+            // >= 6 because other tests running in parallel may have also resolved the same host.
```

Comment:
The test doesn't protect from firing the event more times than needed. Maybe it's worth moving them out of process instead?